packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -0,0 +1,304 @@
+
+  let selectState = useSelectState(props);
+
+  let selectedKeyItem = selectedKey ? selectState.collection.getItem(selectedKey) : undefined;
+  let selectedKeyText = selectedKeyItem ? selectedKeyItem.textValue || selectedKeyItem.rendered as string : undefined;
+  // Maybe don't need to do for defaultSelectedKey? Need to do for selectedKey so that the textfield is properly controlled reflects selectedKey text
+  let defaultSelectedKeyItem = defaultSelectedKey ? selectState.collection.getItem(defaultSelectedKey) : undefined;
+  let defaultSelectedKeyText = defaultSelectedKeyItem ? defaultSelectedKeyItem.textValue || defaultSelectedKeyItem.rendered as string : undefined;
+  // Double check if selectedKey should make textfield value controlled
+  let [value, setValue] = useControlledState(toString(inputValue) || selectedKeyText, toString(defaultInputValue) || defaultSelectedKeyText || '', onInputChange);
+  let lowercaseValue = value.toLowerCase();
+
+  let defaultFilterFn = useMemo(() => (node: Node<T>) => {
+    let scan = 0;
+    let lowercaseNode = node.textValue.toLowerCase();
+    for (let i in lowercaseValue) {
+      if (whitespace.test(lowercaseValue[i])) {
+        continue;
+      }
+      let match = false;
+      for (; scan < lowercaseNode.length && !match; scan++) {
+        if (!whitespace.test(lowercaseNode[scan])) { // if whitespace, move on to next character
+          let compareVal = collator.compare(lowercaseValue[i], lowercaseNode[scan]);
+          if (compareVal !== 0) { // if it didn't match, we can exit
+            return false;
+          } else { // we matched, move to next character in input for comparison
+            match = true;
+          }
+        }
+      }
+    }
+    return true;
+  }, [collator, lowercaseValue]);
+
+  useEffect(() => {
+    if (onFilter) {
+      onFilter(value);
+    }
+  // Having onFilter in the dep array seems to break it
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [value]);
+
+  selectState.collection = useMemo(() => {
+    if (itemsControlled || value === '') {
+      return selectState.collection;
+    }
+    return new FilteredCollection(selectState.collection, defaultFilterFn);
+  }, [selectState.collection, value, itemsControlled, defaultFilterFn]);
+ 
+  useEffect(() => {
+    // Close the menu if it was open but there aren't any items to display
+    if (menuIsOpen && selectState.collection.size === 0) {
+      selectState.close();
+    }
+
+    // Only open the menu if there are items to display and the combobox is focused
+    // Note: doesn't affect controlled isOpen or defaultOpen
+    if (isFocused && menuIsOpen && selectState.collection.size > 0) {
+      selectState.open();
+    }
+
+    // Close the menu if it is supposed to be closed
+    if (!menuIsOpen) {
+      selectState.close();
+    }
+
+    // Maybe change dep array back to selectState? or make it selectState.close?
+  }, [menuIsOpen, selectState.collection, selectState, isFocused]);
```

Comment:
Note that this will NOT make onOpenChange fire if the combobox is controlled isOpen and the user attempts to close the combobox via pressing the combobox button or using the up/down arrow keys. A bit of a departure from what we usually do for controlled states but it eliminates a bunch of "onOpenChange(true) -> onOpenChange(false)" that happen when the user types in the field but there aren't any options to show